### PR TITLE
Modernize compiler flags

### PR DESCRIPTION
**Pull request by galanCA**
## Description

Change `compiler-flags.cmake` to use a common interface library to use for its targets. It also removes the default flags set by CMake allowing for the developers to have full control of the flags added.

It also removes c++11 flags, warning suppression flags, and turns warning into errors for msvc, GCC and Clang.

## Test
- CI Test.
- Test the following [commit](https://github.com/bigladder/penumbra/pull/11/commits/daf9ce0e19c7678e37f0555efaed3d5c3ed33b11) in a [CSE branch](https://github.com/cse-sim/cse/pull/334) by updating penumbra to said commit.

### CODE REVIEW

**Comment by galanCA** [comment]
@nealkruis Deleting the warning suppression flags made a bunch of warnings appear for ubuntu for vendor libraries. My next steps are to suppress those warnings when we include those headers.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -38,7 +35,10 @@ target_compile_options(${PROJECT_NAME} PRIVATE
   -Wall -Wextra -Wpedantic>
```
Comment (nealkruis) [inline src/CMakeLists.txt]: These should probably be removed in favor of the other settings in compiler-flags.cmake.

**Review by nealkruis — CHANGES_REQUESTED** [review]
While we're at it, we should look for non-target-specific properties (e.g., `include_directories`) and replace them with the target-specific versions.

**Review by nealkruis — APPROVED** [review]

**Review by nealkruis — COMMENTED** [review]
Approval pending fixes to libtess2.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -24,21 +17,24 @@ if (${PROJECT_NAME}_STATIC_LIB)
...
   set(CMAKE_MACOSX_RPATH 1)
   add_library(${PROJECT_NAME} SHARED ${library_sources})
-  if(UNIX)
-    set_target_properties(tess2 PROPERTIES COMPILE_FLAGS "-fPIC")
-  endif()
 endif()
 
+target_include_directories(${PROJECT_NAME} PRIVATE
+  ${PROJECT_SOURCE_DIR}/include
+  ${PROJECT_SOURCE_DIR}/src
```
Comment (nealkruis) [inline src/CMakeLists.txt:24]: Shouldn't these be public?

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -24,21 +17,24 @@ if (${PROJECT_NAME}_STATIC_LIB)
...
+  SYSTEM ${PROJECT_SOURCE_DIR}/vendor/glad/include
+  SYSTEM ${PROJECT_SOURCE_DIR}/vendor/glfw/include
+  SYSTEM ${PROJECT_SOURCE_DIR}/vendor/glfw/deps
+  SYSTEM ${PROJECT_SOURCE_DIR}/vendor/libtess2/Include
```
Comment (nealkruis) [inline src/CMakeLists.txt]: Can we make these public on their respective targets?

File: examples/CMakeLists.txt [hunk 4]
```diff
@@ -14,6 +12,7 @@ endif()
 
 add_executable(awning WIN32 MACOSX_BUNDLE awning.cpp ${ICON})
 
+target_include_directories(awning PRIVATE ${PROJECT_SOURCE_DIR}/include)
```
Comment (nealkruis) [inline examples/CMakeLists.txt]: This shouldn't be necessary if it is public on the penumbra target.

File: test/CMakeLists.txt [hunk 5]
```diff
@@ -1,14 +1,17 @@
...
   ${PROJECT_NAME}_test.cpp)
 
 
 add_executable(${PROJECT_NAME}_tests ${test_sources})
 
-target_link_libraries(${PROJECT_NAME}_tests ${PROJECT_NAME} gtest)
+
+target_include_directories(${PROJECT_NAME}_tests PRIVATE
+  ${PROJECT_SOURCE_DIR}/src
+  ${PROJECT_SOURCE_DIR}/include
```
Comment (nealkruis) [inline test/CMakeLists.txt]: These shouldn't be necessary if they are public on the penumbra target.

**Review by nealkruis — CHANGES_REQUESTED** [review]

File: src/CMakeLists.txt (hunk 2 above)
Comment (galanCA) [reply src/CMakeLists.txt:24]: Done, These directories were changed to `PUBLIC`.

**Review by nealkruis — APPROVED** [review]

**Pull request merged**